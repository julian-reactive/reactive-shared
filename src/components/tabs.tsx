import React, { ReactElement, useCallback, useMemo, useState } from 'react'
import map from 'lodash/map'

// Material Components
import Tabs from '@mui/material/Tabs'
import Tab from '@mui/material/Tab'
import Typography from '@mui/material/Typography'
import Box from '@mui/material/Box'

const a11yProps = (index: string): {id: string, 'aria-controls': string} => ({
  id: `simple-tab-${index}`,
  'aria-controls': `simple-tabpanel-${index}`
})

// Interfaces
interface TabPanelProps {
  children: any
  index: number
  value: number
}

export interface AppTabsProps {
  value?: number
  tabsHeader: Array<number | string>
  tabsBody: ReactElement[] | string[]
}

const TabPanel: React.FC<TabPanelProps> = ({ children, value, index, ...other }) => {
  return (
    <div
      role='tabpanel'
      hidden={value !== index}
      id={`simple-tabpanel-${index}`}
      aria-labelledby={`simple-tab-${index}`}
      {...other}
    >
      {value === index && (
        <Box>
          <Typography>{children}</Typography>
        </Box>
      )}
    </div>
  )
}

const AppTabsContainer: React.FC<AppTabsProps> = ({ value = 0, tabsHeader, tabsBody }) => {
  const [tabValue, setTabValue] = useState(value)

  const handleChange = useCallback((event: React.SyntheticEvent, newValue: string) => setTabValue(parseInt(newValue, 10)), [])

  const renderTabs = useMemo(() => {
    return map(tabsHeader, item => (
      <Tab
        key={`tab-${item}`}
        {...a11yProps(item.toString())}
        label={item}
      />
    ))
  }, [tabsHeader])

  const renderTabPanels = useMemo(() => {
    return map(tabsBody, (item, index) => (
      <TabPanel key={index} index={parseInt(index, 10)} value={tabValue}>
        {item}
      </TabPanel>
    ))
  }, [tabsBody, tabValue])

  return (
    <>
      <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
        <Tabs
          value={tabValue}
          onChange={handleChange}
          variant='scrollable'
          scrollButtons='auto'

        >
          {renderTabs}
        </Tabs>
      </Box>
      {renderTabPanels}
    </>
  )
}

export const AppTabs = React.memo(AppTabsContainer)
