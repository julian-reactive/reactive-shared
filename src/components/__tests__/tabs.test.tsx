import React from 'react'

import { render } from '@testing-library/react'
import { AppTabs, AppTabsProps } from '../tabs'

const defaultProps: AppTabsProps = {
  tabsHeader: [1, 2],
  tabsBody: ["Juan Diego", "Juan Davis"]
}

const getRendered = (props = defaultProps) => render(<AppTabs {...props} />)

describe('tabs', () => {
  it('should match snapshot', () => {
    const { container } = getRendered()
    expect(container.firstChild).toMatchSnapshot()
  })
})
