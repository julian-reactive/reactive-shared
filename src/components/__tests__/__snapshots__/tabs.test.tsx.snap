// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tabs should match snapshot 1`] = `
<div
  class="MuiBox-root css-1gsv261"
>
  <div
    class="MuiTabs-root css-1ujnqem-MuiTabs-root"
  >
    <div
      class="MuiTabs-scrollableX MuiTabs-hideScrollbar css-agdkjh-MuiTabs-scrollbarSize"
      style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
    />
    <div
      class="MuiTabs-scroller MuiTabs-hideScrollbar MuiTabs-scrollableX css-69z67c-MuiTabs-scroller"
      style="margin-bottom: 0px;"
    >
      <div
        class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
        role="tablist"
      >
        <button
          aria-controls="simple-tabpanel-1"
          aria-selected="true"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
          id="simple-tab-1"
          role="tab"
          tabindex="0"
          type="button"
        >
          1
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-controls="simple-tabpanel-2"
          aria-selected="false"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
          id="simple-tab-2"
          role="tab"
          tabindex="-1"
          type="button"
        >
          2
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <span
        class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
        style="left: 0px; width: 0px;"
      />
    </div>
  </div>
</div>
`;
